### PR TITLE
RATIS-2149.Do not perform leader election if the current RaftServer has not started yet

### PR DESCRIPTION
**Pull request by 133tosakarin**
see https://issues.apache.org/jira/browse/RATIS-2149

### CODE REVIEW

File: ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderElection.java [hunk 1]
```diff
@@ -234,12 +234,17 @@ LifeCycle.State getCurrentState() {
...
+        LOG.info("{}: skip running since the server is initializing", this);
+        return;
+    }
     if (!lifeCycle.compareAndTransition(STARTING, RUNNING)) {
       final LifeCycle.State state = lifeCycle.getCurrentState();
       LOG.info("{}: skip running since this is already {}", this, state);
       return;
     }
 
+
```
Comment (szetszwo) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderElection.java:247]: Please revert this whitespace change.

File: ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java [hunk 2]
```diff
@@ -1513,6 +1513,10 @@ public CompletableFuture<AppendEntriesReplyProto> appendEntriesAsync(
     }
   }
 
+  public boolean initializing() {
+    return !startComplete.get();
+  }
```
Comment (szetszwo) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java:1518]: Let's check also the `lifeCycle`.  Also, remove `public`.
```java
  boolean isRunning() {
    return startComplete.get() && lifeCycle.getCurrentState() == State.RUNNING;
  }
```

File: ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderElection.java [hunk 3]
```diff
@@ -234,12 +234,17 @@ LifeCycle.State getCurrentState() {
 
   @Override
   public void run() {
+    if (server.initializing()) {
+        LOG.info("{}: skip running since the server is initializing", this);
```
Comment (szetszwo) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderElection.java:238]: Update the message after changing the if-condition.

**Review by szetszwo — COMMENTED** [review]
@133tosakarin , thanks a lot for working on this!  Please see the comments inlined and also below:
```java
diff --git a/ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderElection.java b/ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderElection.java
index da9c513486..aca42d80d8 100644
--- a/ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderElection.java
+++ b/ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderElection.java
@@ -234,6 +234,10 @@ class LeaderElection implements Runnable {
 
   @Override
   public void run() {
+    if (!server.isRunning()) {
+        LOG.info("{}: skip since the server is not running", this);
+        return;
+    }
     if (!lifeCycle.compareAndTransition(STARTING, RUNNING)) {
       final LifeCycle.State state = lifeCycle.getCurrentState();
       LOG.info("{}: skip running since this is already {}", this, state);
diff --git a/ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java b/ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java
index c682de8bfb..79ca22e18d 100644
--- a/ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java
+++ b/ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java
@@ -1465,6 +1465,10 @@ class RaftServerImpl implements RaftServer.Division,
     }
   }
 
+  boolean isRunning() {
+    return startComplete.get() && lifeCycle.getCurrentState() == State.RUNNING;
+  }
+
   @Override
   public CompletableFuture<ReadIndexReplyProto> readIndexAsync(ReadIndexRequestProto request) throws IOException {
     assertLifeCycleState(LifeCycle.States.RUNNING);
```

File: ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderElection.java (hunk 1 above)
Comment (133tosakarin) [reply ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderElection.java:247]: Sorry, I deleted this blank line, but forgot to push it again.

The implementation you provided is very good, I will modify it right away

File: ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java (hunk 2 above)
Comment (133tosakarin) [reply ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java:1518]: Ok, I will refer to this way.

File: ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderElection.java (hunk 3 above)
Comment (133tosakarin) [reply ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderElection.java:238]: ok, thanks your comment

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

**Comment by 133tosakarin** [comment]
Since I can't pass the testImmediatelyRevertedToFollower, I move the if condition to the try block.

**Review by OneSizeFitsQuorum — APPROVED** [review]
LGTM

**Comment by szetszwo** [comment]
`TestRaftWithGrpc` keeps failing.  The failure does not seem related to this.

**Comment by szetszwo** [comment]
Filed RATIS-2151 for the failure of `TestRaftWithGrpc`.

**Comment by szetszwo** [comment]
Retried 20 times but still cannot get a clean build https://github.com/apache/ratis/actions/runs/10699620016/job/29810958383?pr=1143#step:5:735

Since the failure of `TestRaftWithGrpc` is unrelated, let's merge this.

**Pull request merged**